File: clients/ruby/test/property_test.rb
# frozen_string_literal: true

# What holds for every input, rather than for the ones a case happened to pick.
#
# Four things are checked here. A retry schedule never spends more than the policy that produced it
# allows, whichever way the randomness fell. Reading a signature header answers with the one failure
# this gem declares, whatever text reached the endpoint, and never with anything else. A value read
# out of a document the API could answer is written back as the value that was read. And identifiers
# minted in sequence never go back in time.
#
# There is no `hypothesis`-grade tool in Ruby's standard library, and this gem installs nothing at
# runtime, so the search is written here: a fixed seed, a bounded number of draws, and the
# counter-examples worth keeping committed under `regressions/` so they run as ordinary cases on
# every pipeline. A failing draw is one somebody can reproduce by running the suite again rather
# than one that goes away on a retry.

require_relative "test_helper"

module Hook0Test
  class PropertyTest < Minitest::Test
    # What the draws are made from. Fixed, so the suite explores the same inputs everywhere it runs.
    SEED = 20_260_814

    # How many draws each property makes. Bounded, so a pipeline can never be held by one.
    DRAWS = 200

    # How far two sums of the same floats may sit apart before the difference is a defect rather
    # than the order they were added in.
    ROUNDING = 1e-9

    # The bounds a drawn policy is built inside.
    MAX_DRAWN_ATTEMPTS = 64
    MAX_DRAWN_SECONDS = 10.0
    MAX_DRAWN_BUDGET = 60.0

    # Longest header a draw builds.
    MAX_DRAWN_HEADER = 96

    def setup
      @random = Random.new(SEED)
    end

    def test_a_retry_schedule_stays_within_every_bound_of_its_policy
      cases = Hook0Test.corpus("retry_policies") + Array.new(DRAWS) { drawn_policy }

      cases.each do |max_attempts, initial_backoff, max_backoff, max_total_delay, draws|
        policy = Hook0::RetryPolicy.new(
          max_attempts: max_attempts,
          initial_backoff: initial_backoff,
          max_backoff: max_backoff,
          max_total_delay: max_total_delay
        )
        holds_for(policy, draws.map { |drawn| unusable(drawn) })
      end
    end

    def test_reading_a_signature_answers_with_the_one_failure_this_gem_declares
      headers = Hook0Test.corpus("signatures") + Array.new(DRAWS) { drawn_header }

      headers.each do |header|
        begin
          Hook0::Signature.parse(header)
        rescue Hook0::ClientError
          next
        end

        # Parsing answered, so verifying has to answer the same way: a header that reads must not
        # find a way to fail that a caller cannot name.
        begin
          Hook0.verify_webhook_signature_with_current_time(header, "", {}, "secret", 300.0, Time.at(0))
        rescue Hook0::ClientError
          next
        end
      end
    end

    def test_a_generated_type_reads_back_what_it_wrote
      documents = Hook0Test.corpus("documents")
      drawn = documents.flat_map { |document| Array.new(4) { mutated(document) } }

      declared_models.each do |model|
        (documents + drawn).each do |document|
          begin
            read = model.from_json(document)
          rescue Hook0::Runtime::DecodeError
            next
          end

          written = read.to_h

          assert_equal read, model.from_json(written), "a #{model} read out of #{document} does not read back"
          assert_equal written, model.from_json(written).to_h
        end
      end
    end

    def test_reading_a_document_answers_with_the_one_failure_the_runtime_declares
      documents = Hook0Test.corpus("documents").flat_map { |document| Array.new(4) { mutated(document) } }

      declared_models.each do |model|
        documents.each do |document|
          model.from_json(document)
        rescue Hook0::Runtime::DecodeError
          next
        end
      end
    end

    def test_minted_identifiers_carry_a_moment_that_never_goes_back
      moments = Array.new(DRAWS) { Hook0.generate_event_id }.map { |id| id[0, 8] + id[9, 4] }

      assert_equal moments.sort, moments
    end

    private

    # Every class the generator wrote, found by looking at what it wrote.
    #
    # Nothing lists the types here: a schema the document adds joins this suite the moment the
    # generated files carry it.
    def declared_models
      Hook0::Generated.constants.map { |named| Hook0::Generated.const_get(named) }
                      .select { |declared| declared.is_a?(Class) && declared.respond_to?(:from_json) }
                      .sort_by(&:name)
    end

    def holds_for(policy, draws)
      delays = policy.delays(draws)
      budget = [policy.max_total_delay, 0.0].max

      assert_operator policy.attempts, :>=, 1
      assert_operator policy.attempts, :<=, Hook0::RetryPolicy::MAX_ATTEMPTS_CAP
      assert_operator delays.size, :<=, policy.attempts - 1
      assert_operator delays.sum, :<=, budget + ROUNDING

      delays.each_with_index do |delay, index|
        assert_operator delay, :>=, 0.0
        assert_operator delay, :<=, policy.backoff_ceiling(index + 1) + ROUNDING
        assert_operator delay, :<=, [policy.max_backoff, 0.0].max + ROUNDING
      end

      # A schedule never hurries up as it goes: the ceiling of a retry never sits below the one
      # before it.
      ceilings = (1..policy.attempts).map { |retry_number| policy.backoff_ceiling(retry_number) }

      assert_equal ceilings.sort, ceilings
    end

    # A draw that is no draw at all, which has to make the client wait longer rather than less.
    def unusable(drawn)
      case drawn
      when "nan" then Float::NAN
      when "infinity" then Float::INFINITY
      when "-infinity" then -Float::INFINITY
      else drawn
      end
    end

    def drawn_policy
      [
        @random.rand(-4..MAX_DRAWN_ATTEMPTS),
        @random.rand * MAX_DRAWN_SECONDS,
        @random.rand * MAX_DRAWN_SECONDS,
        @random.rand * MAX_DRAWN_BUDGET,
        Array.new(@random.rand(0..8)) { (@random.rand * 2) - 0.5 }
      ]
    end

    # A header built out of the pieces a signature is made of, put together every way a sender that
    # is not Hook0 might put them together.
    def drawn_header
      pieces = %w[t v0 v1 h = , 0 9 zz abc x-event-id 1800000000 -1 " " . { }]
      Array.new(@random.rand(0..MAX_DRAWN_HEADER)) { pieces.sample(random: @random) }.join
    end

    # A document with one of its members taken away, replaced by something of another type, or
    # buried inside something else.
    def mutated(document)
      return document unless document.is_a?(Hash) && !document.empty?

      key = document.keys.sample(random: @random)
      case @random.rand(0..3)
      when 0 then document.reject { |name, _| name == key }
      when 1 then document.merge(key => @random.rand(0..1_000))
      when 2 then document.merge(key => [document[key]])
      else document.merge(key => nil)
      end
    end
  end
end

File: clients/go/property_test.go
// What holds for every input, rather than for the ones a case happened to pick.
//
// Three things are checked here. A retry schedule never spends more than the policy that produced it
// allows, whichever way the randomness fell. Reading a signature header answers with one of the
// reasons this package declares, whatever text reached the endpoint, and never with a panic. And a
// value written back the way the API reads it is the value that was read.
//
// Each of them is a fuzz target with a corpus committed beside it, so the counterexamples that were
// worth finding run as ordinary cases on every pipeline, and the search itself runs under a
// deadline rather than until somebody stops it.

package hook0_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"net/http"
	"testing"
	"time"

	hook0 "github.com/hook0/hook0-go/v2"
	"github.com/hook0/hook0-go/v2/generated"
)

// refusals are every reason this package declines to verify a webhook. A refusal that is none of
// them is one a caller cannot tell apart from another.
var refusals = []error{
	hook0.ErrSignatureUnreadable,
	hook0.ErrHeaderNotDelivered,
	hook0.ErrSignatureMismatch,
	hook0.ErrSignatureOutsideTolerance,
}

func FuzzRetryScheduleStaysWithinEveryBound(f *testing.F) {
	f.Fuzz(func(
		t *testing.T,
		maxAttempts int,
		initialBackoff int64,
		maxBackoff int64,
		maxTotalDelay int64,
		firstDraw float64,
		secondDraw float64,
	) {
		policy := hook0.RetryPolicy{
			MaxAttempts:    maxAttempts,
			InitialBackoff: time.Duration(initialBackoff),
			MaxBackoff:     time.Duration(maxBackoff),
			MaxTotalDelay:  time.Duration(maxTotalDelay),
		}
		delays := policy.Delays([]float64{firstDraw, secondDraw})

		attempts := policy.Attempts()
		if attempts < 1 || attempts > hook0.MaxAttemptsCap {
			t.Fatalf("a policy of %d attempts makes %d of them", maxAttempts, attempts)
		}
		if len(delays) > attempts-1 {
			t.Fatalf("%d attempts are spaced out by %d delays", attempts, len(delays))
		}

		budget := max(policy.MaxTotalDelay, 0)
		ceiling := max(policy.MaxBackoff, 0)

		// Written as what is left of the budget rather than as a running sum, so the check itself
		// cannot carry two durations past what one can hold.
		var spent time.Duration
		for index, delay := range delays {
			if delay < 0 {
				t.Fatalf("retry %d waits for %s", index+1, delay)
			}
			if delay > policy.BackoffCeiling(index+1) {
				t.Fatalf("retry %d waits %s, above its ceiling of %s", index+1, delay, policy.BackoffCeiling(index+1))
			}
			if delay > ceiling {
				t.Fatalf("retry %d waits %s, above the %s no delay exceeds", index+1, delay, ceiling)
			}
			if delay > budget-spent {
				t.Fatalf("the schedule spends more than the %s budget its delays share", budget)
			}
			spent += delay
		}

		// A schedule never hurries up as it goes: the ceiling of a retry never sits below the one
		// before it.
		previous := policy.BackoffCeiling(1)
		for retry := 2; retry <= attempts; retry++ {
			current := policy.BackoffCeiling(retry)
			if current < previous {
				t.Fatalf("retry %d has a ceiling of %s where retry %d had %s", retry, current, retry-1, previous)
			}
			previous = current
		}
	})
}

func FuzzReadingASignatureAnswersOneOfTheDeclaredRefusals(f *testing.F) {
	f.Fuzz(func(t *testing.T, header string, payload []byte) {
		// Parsing on its own must answer the same way: everything below it depends on it not
		// answering something a caller cannot name.
		if _, err := hook0.ParseSignature(header); err != nil && !errors.Is(err, hook0.ErrSignatureUnreadable) {
			t.Fatalf("reading %q answered %v, which is not a reason this package declares", header, err)
		}

		err := hook0.VerifyWebhookSignatureAt(
			header,
			payload,
			deliveredHeaders(),
			subscriptionSecret,
			tolerance,
			signedMoment(),
		)
		if err == nil {
			return
		}
		for _, declared := range refusals {
			if errors.Is(err, declared) {
				return
			}
		}
		t.Fatalf("verifying %q answered %v, which is not a reason this package declares", header, err)
	})
}

func FuzzAGeneratedTypeWritesBackWhatItRead(f *testing.F) {
	f.Fuzz(func(t *testing.T, document []byte) {
		// Two types, picked for the shapes they cover between them: the error contract's own, whose
		// discriminant is a closed list and whose last member is a value the document does not
		// describe, and the deepest one the API declares, which nests objects, an identifier, and
		// members it does not require.
		roundTrips[generated.Problem](t, document)
		roundTrips[generated.ApplicationInfo](t, document)
	})
}

// roundTrips checks that a document this type could read is written back as the document it read.
//
// What is held against what is not the bytes that arrived but the bytes the type writes: a document
// may spell a number or an escape more ways than one, and the question is whether the value
// survives, not whether the sender's spelling does.
func roundTrips[Value any](t *testing.T, document []byte) {
	t.Helper()

	var read Value
	if err := json.Unmarshal(document, &read); err != nil {
		return
	}

	written, err := json.Marshal(read)
	if err != nil {
		t.Fatalf("a %T read out of %q cannot be written back: %v", read, document, err)
	}

	var again Value
	if err := json.Unmarshal(written, &again); err != nil {
		t.Fatalf("a %T writes back %q, which it cannot read: %v", read, written, err)
	}

	rewritten, err := json.Marshal(again)
	if err != nil {
		t.Fatalf("a %T read out of its own %q cannot be written back: %v", read, written, err)
	}
	if !bytes.Equal(written, rewritten) {
		t.Fatalf("a %T read out of %q writes back %q, then %q", read, document, written, rewritten)
	}
}

func TestTheDefaultScheduleDoublesUpToItsCeiling(t *testing.T) {
	policy := hook0.DefaultRetryPolicy()

	for _, expected := range []struct {
		retry   int
		ceiling time.Duration
	}{
		{1, 100 * time.Millisecond},
		{2, 200 * time.Millisecond},
		{3, 400 * time.Millisecond},
		{6, 2 * time.Second},
		{16, 2 * time.Second},
	} {
		if reached := policy.BackoffCeiling(expected.retry); reached != expected.ceiling {
			t.Errorf("retry %d has a ceiling of %s, not %s", expected.retry, reached, expected.ceiling)
		}
	}

	// A source of randomness that gives nothing asks for the whole ceiling, which is what makes the
	// budget the thing that cuts the schedule short.
	delays := policy.Delays(nil)
	if len(delays) != 3 {
		t.Fatalf("four attempts are spaced out by %d delays", len(delays))
	}

	var spent time.Duration
	for _, delay := range delays {
		spent += delay
	}
	if spent > policy.MaxTotalDelay {
		t.Errorf("the schedule spends %s of a %s budget", spent, policy.MaxTotalDelay)
	}
}

func TestADisabledPolicyWaitsForNothing(t *testing.T) {
	policy := hook0.DisabledRetryPolicy()

	if attempts := policy.Attempts(); attempts != 1 {
		t.Errorf("a disabled policy makes %d attempts", attempts)
	}
	if delays := policy.Delays([]float64{1, 1, 1}); len(delays) != 0 {
		t.Errorf("a disabled policy waits %v between its attempts", delays)
	}
}

func TestAPopulatedValueSurvivesBeingWrittenAndReadBack(t *testing.T) {
	document, err := json.Marshal(anApplication())
	if err != nil {
		t.Fatalf("the case cannot write the document it starts from: %v", err)
	}

	var read generated.ApplicationInfo
	if err := json.Unmarshal(document, &read); err != nil {
		t.Fatalf("the document the API answers cannot be read: %v", err)
	}

	written, err := json.Marshal(read)
	if err != nil {
		t.Fatalf("what was read cannot be written back: %v", err)
	}

	var again generated.ApplicationInfo
	if err := json.Unmarshal(written, &again); err != nil {
		t.Fatalf("what was written back cannot be read: %v", err)
	}

	if again.ApplicationId != read.ApplicationId ||
		again.Name != read.Name ||
		again.OrganizationId != read.OrganizationId ||
		again.Quotas.EventsPerDayLimit != read.Quotas.EventsPerDayLimit ||
		again.OnboardingSteps.Event != read.OnboardingSteps.Event {
		t.Errorf("the value changed on the way round: %+v became %+v", read, again)
	}
}

func TestAMemberTheDocumentDoesNotRequireIsAbsentRatherThanZero(t *testing.T) {
	var read generated.Problem
	if err := json.Unmarshal([]byte(`{"id":"NotFound","title":"t","detail":"d","status":404,"type":"u"}`), &read); err != nil {
		t.Fatalf("a problem document cannot be read: %v", err)
	}

	written, err := json.Marshal(read)
	if err != nil {
		t.Fatalf("a problem cannot be written back: %v", err)
	}
	if bytes.Contains(written, []byte(`"validation"`)) {
		t.Errorf("a member the API did not answer was written back anyway: %s", written)
	}
}

func TestVerifyingAgainstTheCurrentMomentRefusesAStaleDelivery(t *testing.T) {
	// The fixed vectors were signed long enough ago that no clock puts them inside the window.
	err := hook0.VerifyWebhookSignature(
		"t="+"1800000000"+",v0="+bodyCode,
		[]byte(signedPayload),
		http.Header{},
		subscriptionSecret,
		tolerance,
	)
	if !errors.Is(err, hook0.ErrSignatureOutsideTolerance) {
		t.Errorf("a delivery signed years ago was read as %v", err)
	}
}
